### ISSUE TITLE
[Codegen] Spec not adhered to wrt. relative Server URLs

### ISSUE THREAD

**Issue by ahri** [swagger-api/swagger-codegen#7960]
##### Description

The spec for a [Server Object](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.0.md#server-object) states that the URL is "A URL to the target host. This URL supports Server Variables **and MAY be relative**, to indicate that the host location is relative to the location where the OpenAPI document is being served" (emphasis mine).

The generator does not support relative URLs:

```
[Thread-0] WARN io.swagger.codegen.utils.URLPathUtil - Not valid URL: api
java.net.MalformedURLException: no protocol: api
	at java.base/java.net.URL.<init>(URL.java:627)
	at java.base/java.net.URL.<init>(URL.java:523)
	at java.base/java.net.URL.<init>(URL.java:470)
	at io.swagger.codegen.utils.URLPathUtil.getServerURL(URLPathUtil.java:30)
	at io.swagger.codegen.DefaultGenerator.buildSupportFileBundle(DefaultGenerator.java:673)
	at io.swagger.codegen.DefaultGenerator.generate(DefaultGenerator.java:734)
	at io.swagger.codegen.cmd.Generate.run(Generate.java:331)
	at java.base/java.lang.Thread.run(Thread.java:844)
```

##### Swagger-codegen version

3.0.0-SNAPSHOT (swagger-codegen-cli-3.0.0-20180328.231509-53.jar)

##### Swagger declaration file content or url

```json
"servers":[{"url":"api"}], ...
```

##### Command line used for generation

```bash
$ java -jar ~/swagger-codegen-cli-3.0.0-20180328.231509-53.jar generate -i http://localhost:48781/v1/openapi.json -l java -o /tmp/test/
```

##### Steps to reproduce

Specify a relative URL for a server object definition and provide the codegen tool with a URL to the OpenAPI JSON file, e.g. `http://foo.com/openapi.json` specifying

```json
"servers":[{"url":"api"}], ...
```

and expecting the URL to be expanded to `http://foo.com/api`.

##### Related issues/PRs

Did not find related issues.

##### Suggest a fix/enhancement

My reading of the spec is that since a relative URL is a valid value for the `url` field, it should be used, along with the input file provided to the codegen (i.e. via the `-i` switch) to derive a new URL.

This will allow OpenAPI files to be served from behind proxies/load balancers where the actual server cannot know its public address with certainty (`X-Forwarded-Host` can be wrong).

**Comment by tkingsbury42** [comment]
I'm running into this issue as well with the latest 3.0.0 snapshot. If I switch back to the latest 2.4.0 snapshot, the problem goes away.

**Comment by dementati** [comment]
This issue still appears to be present in 3.0.0.

**Comment by deg0nz** [comment]
I'm running into the same error.
My API server is behind an NGINX and for `https://myserver.example.com/api/v1` I get the error for a `MalformedURLException` for `///api/v1`.

Swagger API on the server is version 2.3.0 and the relative URl on the swagger.json is `/api/v1`
I'm using 3.0.3 of the codegen.

I worked around it by hard coding the entire server URL into line 32 of `URLPathUtil.java` as @ahri's fix mentioned above didn't work for me.

**Comment by mrjoe7** [comment]
I noticed that when `host` and `schemes` are omitted in `swagger: '2.0'` definition then `basePathWithoutHost` in templates is blank string. It's present in all codegen versions 3.0.x (tested up to 3.0.3).

**Comment by HugoMario** [comment]
this issue has been fixed with PR #9047, so going to close this issue but please let me know if you find something wrong and i'll work on it.

**Comment by OldIMP** [comment]
R u guys planning to merge this fix to the 2.x master branch any time soon? Thx for the great work!
